### PR TITLE
ci: build kokkos tools with kokkos

### PR DESCRIPTION
**Pull request by romintomasetti**
As promised in #182, here is a better version of the CI, that builds against `OpenMP`, `Cuda` and `HIP` backends.

Things to discuss:
- [x] Do we keep the "simple CI" version? -> Yes, it is important to keep a "standalone" build.
- [x] Do you agree with the CMake presets based stuff?

### CODE REVIEW

File: kokkos.presets.json [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        "patch" : 0
+    },
+    "configurePresets" : [
+        {
+            "name"          : "default",
+            "displayName"   : "default",
+            "binaryDir"     : "${sourceDir}/build-with-${presetName}",
+            "cacheVariables" : {
+                "CMAKE_EXPORT_COMPILE_COMMANDS" : {"type" : "BOOL"  , "value" : "ON"},
+                "CMAKE_BUILD_TYPE"              : {"type" : "STRING", "value" : "RELEASE"},
```
Comment (dalg24) [inline kokkos.presets.json]: Isn't it spelled "Release"?

File: kokkos.presets.json [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        },
+        {
+            "name"        : "GNU-Cuda",
+            "displayName" : "GNU-Cuda",
+            "inherits"    : "default",
+            "cacheVariables" : {
+                "Kokkos_ENABLE_CUDA"        : {"type" : "BOOL"  , "value" : "ON"},
+                "Kokkos_ENABLE_CUDA_LAMBDA" : {"type" : "BOOL"  , "value" : "ON"},
+                "Kokkos_ARCH_VOLTA70"       : {"type" : "BOOL"  , "value" : "ON"},
+                "CMAKE_CXX_COMPILER"        : {"type" : "STRING", "value" : "$penv{PWD}/bin/nvcc_wrapper"}
```
Comment (dalg24) [inline kokkos.presets.json]: Why are you using `penv${PWD}` instead of just `env${PWD}`?

**Comment by crtrott** [comment]
I think we should keep the "simple CI" since it tests a build without Kokkos Core present which I'd like to keep going for simple tools.

Otherwise the preset stuff looks ok to me, any other thoughts?

File: kokkos.presets.json [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    "version"   : 3,
+    "cmakeMinimumRequired" : {
+        "major" : 3,
+        "minor" : 22,
+        "patch" : 0
+    },
+    "configurePresets" : [
+        {
+            "name"          : "default",
+            "displayName"   : "default",
```
Comment (dalg24) [inline kokkos.presets.json]: What's the point of using `displayName` if we are just echoing the name of the preset?

File: kokkos.presets.json [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            "binaryDir"     : "${sourceDir}/build-with-${presetName}",
+            "cacheVariables" : {
+                "CMAKE_EXPORT_COMPILE_COMMANDS" : {"type" : "BOOL"  , "value" : "ON"},
+                "CMAKE_BUILD_TYPE"              : {"type" : "STRING", "value" : "RELEASE"},
+                "CMAKE_CXX_STANDARD"            : {"type" : "STRING", "value" : "17"},
+                "CMAKE_CXX_EXTENSIONS"          : {"type" : "BOOL"  , "value" : "OFF"},
+                "BUILD_SHARED_LIBS"             : {"type" : "BOOL"  , "value" : "ON"}
```
Comment (dalg24) [inline kokkos.presets.json]: Does it matter?

File: kokkos.presets.json [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            }
+        },
+        {
+            "name"        : "GNU-OpenMP",
+            "displayName" : "GNU-OpenMP",
+            "inherits"    : "default",
+            "cacheVariables" : {
+                "Kokkos_ENABLE_OPENMP" : {"type" : "BOOL"  , "value" : "ON"}
```
Comment (dalg24) [inline kokkos.presets.json]: I am not a big fan of the verbose syntax when we can get away with `"Kokkos_ENABLE_OPENMP": "ON"` but not blocking if others feel strongly about it.

File: kokkos.presets.json [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+            "cacheVariables" : {
+                "CMAKE_EXPORT_COMPILE_COMMANDS" : {"type" : "BOOL"  , "value" : "ON"},
+                "CMAKE_BUILD_TYPE"              : {"type" : "STRING", "value" : "RELEASE"},
+                "CMAKE_CXX_STANDARD"            : {"type" : "STRING", "value" : "17"},
+                "CMAKE_CXX_EXTENSIONS"          : {"type" : "BOOL"  , "value" : "OFF"},
+                "BUILD_SHARED_LIBS"             : {"type" : "BOOL"  , "value" : "ON"}
+            }
+        },
+        {
+            "name"        : "GNU-OpenMP",
```
Comment (dalg24) [inline kokkos.presets.json]: Why the "GNU-" prefix?

**Review by dalg24 — COMMENTED** [review]
* Definitely keep the old build, the standalone build without Kokkos should still be considered as the default way to build Kokkos Tools
* No objection to the CMake presets

**Comment by crtrott** [comment]
@romintomasetti btw, if you want to get more formally involved with Kokkos Tools we are happy to do that :-)

File: kokkos.presets.json (hunk 1 above)
Comment (romintomasetti) [reply kokkos.presets.json]: Indeed. Done.

File: kokkos.presets.json (hunk 2 above)
Comment (romintomasetti) [reply kokkos.presets.json]: In this case, it seems that using `${sourceDir}/bin/nvcc_wrapper` is cleaner than relying on some environment variable.

File: kokkos.presets.json (hunk 3 above)
Comment (romintomasetti) [reply kokkos.presets.json]: Indeed. I removed `displayName` :smile:

File: kokkos.presets.json (hunk 4 above)
Comment (romintomasetti) [reply kokkos.presets.json]: I don't know. By default, this is `OFF` in `Kokkos`, see https://github.com/kokkos/kokkos/blob/55bbd9f175307a2ad42d205a3cebb5a32698545b/CMakeLists.txt#L225.

I don't have any opinion on that in this case. If you want, I can remove the `BUILD_SHARED_LIBS` parameter.

File: kokkos.presets.json (hunk 6 above)
Comment (romintomasetti) [reply kokkos.presets.json]: I think the rationale is that if you compile with e.g. `clang` it would become `clang-OpenMP`. Fine to do what you prefer though.

**Comment by romintomasetti** [comment]
> @romintomasetti btw, if you want to get more formally involved with Kokkos Tools we are happy to do that :-)

@crtrott I'd be more than happy to participate in the adventure!

File: kokkos.presets.json [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            "cacheVariables" : {
+                "Kokkos_ENABLE_OPENMP" : {"type" : "BOOL"  , "value" : "ON"}
+            }
+        },
+        {
+            "name"        : "GNU-Cuda",
+            "inherits"    : "default",
+            "cacheVariables" : {
+                "Kokkos_ENABLE_CUDA"        : {"type" : "BOOL"  , "value" : "ON"},
+                "Kokkos_ENABLE_CUDA_LAMBDA" : {"type" : "BOOL"  , "value" : "ON"},
```
Comment (dalg24) [inline kokkos.presets.json]: If not actually necessary for this PR I would recommend to drop since the option defaults to ON in 4.0 and there is discussion about deprecating it in 4.1 (see kokkos/kokkos#5964)

File: kokkos.presets.json (hunk 6 above)
Comment (dalg24) [reply kokkos.presets.json]: I'd rather drop it and reintroduce later if necessary.

File: CMakePresets.json [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    "configurePresets" : [
+        {
+            "name"          : "default",
+            "binaryDir"     : "${sourceDir}/build-with-${presetName}",
+            "cacheVariables" : {
+                "CMAKE_EXPORT_COMPILE_COMMANDS"     : {"type" : "BOOL"  , "value" : "ON"},
+                "CMAKE_BUILD_TYPE"                  : {"type" : "STRING", "value" : "Release"},
+                "CMAKE_CXX_STANDARD"                : {"type" : "STRING", "value" : "17"},
+                "KokkosTools_REUSE_KOKKOS_COMPILER" : {"type" : "BOOL"  , "value" : "ON"},
+                "Kokkos_DIR"                        : {"type" : "STRING", "value" : "$penv{Kokkos_DIR}"}
```
Comment (dalg24) [inline CMakePresets.json]: This is not the right option to help find Kokkos.  I'd expect one should use `Kokkos_ROOT`.
I suspect Kokkos is being found because CMake actually parses the environment for `<PackageName>_DIR`.
Either drop it or use Kokkos_ROOT.

**Review by dalg24 — COMMENTED** [review]
I think we are getting close

File: kokkos.presets.json (hunk 6 above)
Comment (romintomasetti) [reply kokkos.presets.json]: Agreed, and done :heavy_check_mark:

File: kokkos.presets.json (hunk 7 above)
Comment (romintomasetti) [reply kokkos.presets.json]: Ok, removed! Thanks for the pointer ;)

File: CMakePresets.json (hunk 8 above)
Comment (romintomasetti) [reply CMakePresets.json]: I think I got inspiration from https://github.com/kokkos/kokkos-tools/blob/3855d8f6186b600b97ae81cd1d9417ad8e34c776/example/CMakeLists.txt#L1-L10, but you're right!

Comment (masterleinad) [reply CMakePresets.json]: FWIW, it's totally fine to set `Kokkos_DIR` pointing to the directory holding `KokkosConfig.cmake`.

File: CMakeLists.txt [hunk 9]
```diff
@@ -16,6 +16,7 @@ include(cmake/configure_tpls.cmake)
 
 # Set policies
 cmake_policy(SET CMP0111 NEW) # error if library not found
+cmake_policy(SET CMP0074 NEW) # use <PackageName>_ROOT
```
Comment (dalg24) [inline CMakeLists.txt]:
```suggestion
```
No that is unnecessary with the cmake minimum required version >= 3.12

File: CMakePresets.json [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    "configurePresets" : [
+        {
+            "name"          : "default",
+            "binaryDir"     : "${sourceDir}/build-with-${presetName}",
+            "cacheVariables" : {
+                "CMAKE_EXPORT_COMPILE_COMMANDS"     : {"type" : "BOOL"  , "value" : "ON"},
+                "CMAKE_BUILD_TYPE"                  : {"type" : "STRING", "value" : "Release"},
+                "CMAKE_CXX_STANDARD"                : {"type" : "STRING", "value" : "17"},
+                "KokkosTools_REUSE_KOKKOS_COMPILER" : {"type" : "BOOL"  , "value" : "ON"},
+                "Kokkos_ROOT"                       : {"type" : "STRING", "value" : "$penv{Kokkos_ROOT}"}
```
Comment (dalg24) [inline CMakePresets.json]: Not blocking but re-iterating that I don't think this is a legitimate use of $penv{...} I think $env{...} should be preferred.
Feel free to ignore but if you change make sure you tackle all occurrences.

File: CMakeLists.txt (hunk 9 above)
Comment (romintomasetti) [reply CMakeLists.txt]: I was still having the warning in the CI, so I wanted to silence it. But fine if you want to remove the policy and accept the console warning ;)

File: CMakePresets.json (hunk 10 above)
Comment (romintomasetti) [reply CMakePresets.json]: Changed to `env`

**Comment by romintomasetti** [comment]
@dalg24 Seems that there is an issue with the "space time stack" profiler... https://github.com/kokkos/kokkos-tools/actions/runs/4864236304/jobs/8672999895

Any idea?

### Edit

It seems `KokkosTools::SpaceTimeStack::kokkosp_init_library` is never called, and so the `global_state` is never initialized.

### Edit

I've added several prints in `Kokkos` and I think I found the issue.

In the function `Kokkos::Impl::OpenMPInternal::resize_thread_data`, that is called before the `KokkosTools::SpaceTimeStack::kokkosp_init_library` is called, it will do `ptr = space.allocate(alloc_bytes);`, that will try to call the space and time stack `KokkosTools::SpaceTimeStack::State::allocate` before `KokkosTools::SpaceTimeStack::kokkosp_init_library` is called.

```bash
Kokkos::Tools::Impl::InitializationStatus Kokkos::Tools::Impl::parse_environment_variables(Kokkos::Tools::InitArguments&)
void {anonymous}::initialize_internal(const Kokkos::InitializationSettings&)
void {anonymous}::initialize_internal(const Kokkos::InitializationSettings&): pre_initialize_internal
void {anonymous}::pre_initialize_internal(const Kokkos::InitializationSettings&)
void {anonymous}::initialize_internal(const Kokkos::InitializationSettings&): initialize_backends
void {anonymous}::initialize_backends(const Kokkos::InitializationSettings&): start
void Kokkos::Impl::ExecSpaceManager::initialize_spaces(const Kokkos::InitializationSettings&)
void Kokkos::Impl::ExecSpaceManager::initialize_spaces(const Kokkos::InitializationSettings&): 050_OpenMP
void Kokkos::Impl::OpenMPInternal::initialize(int): start
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false

void Kokkos::Impl::OpenMPInternal::initialize(int): instantiate OpenMP memory space
void Kokkos::Impl::OpenMPInternal::initialize(int): querying current max threads
void Kokkos::Impl::OpenMPInternal::initialize(int): current max threads 16
void Kokkos::Impl::OpenMPInternal::initialize(int): setting up thread local
void Kokkos::Impl::OpenMPInternal::initialize(int): getting OpenMPInternal singleton
void Kokkos::Impl::OpenMPInternal::initialize(int): resize thread data
void Kokkos::Impl::OpenMPInternal::resize_thread_data(size_t, size_t, size_t, size_t): start
void Kokkos::Impl::OpenMPInternal::resize_thread_data(size_t, size_t, size_t, size_t): rank 0 start
void Kokkos::Impl::OpenMPInternal::resize_thread_data(size_t, size_t, size_t, size_t): rank 0 allocate 28800
void* Kokkos::HostSpace::allocate(size_t) const: allocating 28800
void* Kokkos::HostSpace::allocate(const char*, size_t, size_t) const: allocating 28800
void* Kokkos::HostSpace::impl_allocate(const char*, size_t, size_t, Kokkos::Tools::SpaceHandle) const: allocating 28800
void* Kokkos::HostSpace::impl_allocate(const char*, size_t, size_t, Kokkos::Tools::SpaceHandle) const: profile library loaded so calling Kokkos::Profiling::allocateData(...)
void KokkosTools::SpaceTimeStack::State::allocate(KokkosTools::SpaceTimeStack::Space, const char*, const void*, uint64_t): allocating [unlabeled] in space 0(HOST)
[24f4e59e98fd:16590] *** Process received signal ***
[24f4e59e98fd:16590] Signal: Segmentation fault (11)
[24f4e59e98fd:16590] Signal code: Address not mapped (1)
[24f4e59e98fd:16590] Failing at address: 0xa0
```

### Edit

1. I guess all `Kokkos Tools` functions that rely on a global state should first verify that the global state is in a valid state before trying to call it.

2. I don't know how we should fix this.

**Comment by crtrott** [comment]
This sounds like an issue on the Kokkos Core side, I don't think Kokkos should call any callback before the init thing is called

**Comment by romintomasetti** [comment]
> This sounds like an issue on the Kokkos Core side, I don't think Kokkos should call any callback before the init thing is called

@crtrott Indeed. How do you think we should proceed to fix this? :ambulance:

**Comment by masterleinad** [comment]
```diff
diff --git a/core/src/impl/Kokkos_Core.cpp b/core/src/impl/Kokkos_Core.cpp
index aec334261..a7f91203e 100644
--- a/core/src/impl/Kokkos_Core.cpp
+++ b/core/src/impl/Kokkos_Core.cpp
@@ -780,8 +780,14 @@ void post_initialize_internal(const Kokkos::InitializationSettings& settings) {
 }
 
 void initialize_internal(const Kokkos::InitializationSettings& settings) {
+  // Tools should only be used after the backends have been initialized (because
+  // they have not been initialized at that point) but if they have been set via
+  // set_callbacks they need to be present before calling post_initialize so
+  // that they get initialized as well.
+  Kokkos::Tools::Experimental::pause_tools();
   pre_initialize_internal(settings);
   initialize_backends(settings);
+  Kokkos::Tools::Experimental::resume_tools();
   post_initialize_internal(settings);
 }
```
should be enough to fix this.

**Comment by romintomasetti** [comment]
@masterleinad Will you submit a PR?

Edit: Could the solution be that tools are initialized implicitly the first time they are called? Or should they simply be initialized at the very beginning of kokkos initialisation?

**Comment by masterleinad** [comment]
> Edit: Could the solution be that tools are initialized implicitly the first time they are called? Or should they simply be initialized at the very beginning of kokkos initialisation?

That makes it even more difficult to argue about the initialization of `Kokkos Toosl`.
Instead of the `pause_tools`/`resume_tools` solution above we could also consider initializing the backends after the tools but that messes with the `Legion` use case and might therefore not be an option.

**Comment by romintomasetti** [comment]
> > Edit: Could the solution be that tools are initialized implicitly the first time they are called? Or should they simply be initialized at the very beginning of kokkos initialisation?
> 
> That makes it even more difficult to argue about the initialization of `Kokkos Toosl`. Instead of the `pause_tools`/`resume_tools` solution above we could also consider initializing the backends after the tools but that messes with the `Legion` use case and might therefore not be an option.

What do you suggest then? Should we go for the `pause/resume` option for now and open an issue on both Kokkos and Kokkos Tools to try improve the initialization schemes?

**Comment by masterleinad** [comment]
> What do you suggest then? Should we go for the `pause/resume` option for now and open an issue on both Kokkos and Kokkos Tools to try improve the initialization schemes?

See Kokkos/kokkos#6114.

**Comment by romintomasetti** [comment]
I'm working on this again since https://github.com/kokkos/kokkos/pull/6114 has been merged :)

File: README.md [hunk 11]
```diff
@@ -1,3 +1,7 @@
+![Standalone](https://github.com/kokkos/kokkos-tools/actions/workflows/build-standalone.yml/badge.svg)
+![Kokkos@develop](https://github.com/kokkos/kokkos-tools/actions/workflows/build-with-kokkos.yml/badge.svg)
+![clang-format](https://github.com/kokkos/kokkos-tools/actions/workflows/clang-format-check.yml/badge.svg)
+
```
Comment (dalg24) [inline README.md]: I am not a big fan of badges.  I would remove them but I am not blocking if others think otherwise.

File: .github/workflows/clang-format-check.yml [hunk 12]
```diff
@@ -5,7 +5,7 @@ jobs:
     name: Formatting Check
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v3
```
Comment (dalg24) [inline .github/workflows/clang-format-check.yml:8]: Next time please avoid unrelated changes.  (not asking you to revert)

File: CMakePresets.json [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    },
+    "configurePresets" : [
+        {
+            "name"          : "default",
+            "binaryDir"     : "${sourceDir}/build-with-${presetName}",
+            "cacheVariables" : {
+                "CMAKE_EXPORT_COMPILE_COMMANDS"     : {"type" : "BOOL"  , "value" : "ON"},
+                "CMAKE_BUILD_TYPE"                  : {"type" : "STRING", "value" : "Release"},
+                "CMAKE_CXX_STANDARD"                : {"type" : "STRING", "value" : "17"},
+                "KokkosTools_REUSE_KOKKOS_COMPILER" : {"type" : "BOOL"  , "value" : "ON"},
```
Comment (dalg24) [inline CMakePresets.json]: Where is this used?

File: CMakePresets.json [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+                "KokkosTools_ENABLE_EXAMPLES"       : {"type" : "BOOL"  , "value" : "ON"},
+                "KokkosTools_ENABLE_SINGLE"         : {"type" : "BOOL"  , "value" : "ON"},
+                "KokkosTools_ENABLE_MPI"            : {"type" : "BOOL"  , "value" : "ON"},
+                "Kokkos_ROOT"                       : {"type" : "STRING", "value" : "$env{Kokkos_ROOT}"}
```
Comment (dalg24) [inline CMakePresets.json]:
```suggestion
```
This environment variable is used in the normal search procedure, why would we do that?

File: README.md (hunk 11 above)
Comment (romintomasetti) [reply README.md]: No problem, I'll revert this ;) I was primarily just a mean to trigger the pipeline again (since https://github.com/kokkos/kokkos/pull/6114 was merged in the Kokkos develop branch, I wanted to check all would be fine).

File: .github/workflows/clang-format-check.yml (hunk 12 above)
Comment (romintomasetti) [reply .github/workflows/clang-format-check.yml:8]: Ok.

File: CMakePresets.json (hunk 13 above)
Comment (romintomasetti) [reply CMakePresets.json]: Sorry, it is unclear what you're pointing to. Are you referring to `KokkosTools_REUSE_KOKKOS_COMPILER`?

File: CMakePresets.json (hunk 14 above)
Comment (romintomasetti) [reply CMakePresets.json]: Indeed, weird that I also needed to repeat `Kokkos_ROOT` in the cache variables. Let's remove it and see :)

File: CMakePresets.json (hunk 13 above)
Comment (dalg24) [reply CMakePresets.json]: Yes

Comment (romintomasetti) [reply CMakePresets.json]: Ok :)

The idea of the PR is to add a CI for joint `Kokkos` / `Kokkos Tools` builds. `KokkosTools_REUSE_KOKKOS_COMPILER` is a `Kokkos Tools` (advanced) option that forces to reuse the `Kokkos` compiler (e.g. `nvcc_wrapper` for Cuda I guess).

I thought it would be nice to ensure that during a joint build, both projects are on the same page for the compiler.

What's your opinion on that?

File: CMakePresets.json [hunk 15]
```diff
@@ -0,0 +1,85 @@
+{
+    "version"   : 3,
+    "cmakeMinimumRequired" : {
+        "major" : 3,
+        "minor" : 22,
```
Comment (masterleinad) [inline CMakePresets.json]: Why 22? Isn't the minimum we require 3.16?

File: kokkos.presets.json [hunk 16]
```diff
@@ -0,0 +1,63 @@
+{
+    "version"   : 3,
+    "cmakeMinimumRequired" : {
+        "major" : 3,
+        "minor" : 22,
```
Comment (masterleinad) [inline kokkos.presets.json]: Again 3.16 is the minimum.

File: kokkos.presets.json (hunk 4 above)
Comment (masterleinad) [reply kokkos.presets.json]: I think it's a good idea to only set required CMake options for `Kokkos` since that likely matches more closely a usual user configuration.

File: CMakePresets.json (hunk 15 above)
Comment (romintomasetti) [reply CMakePresets.json]: I removed `cmakeMinimumRequired` from the presets altogether :checkered_flag:

File: kokkos.presets.json (hunk 16 above)
Comment (romintomasetti) [reply kokkos.presets.json]: I removed cmakeMinimumRequired from the presets altogether :checkered_flag:

File: CMakePresets.json (hunk 13 above)
Comment (dalg24) [reply CMakePresets.json]: Oh right.  I had grepped for it and must have misspelled it because now I can see it does exist.

File: kokkos.presets.json (hunk 4 above)
Comment (romintomasetti) [reply kokkos.presets.json]: What do you want me to do exactly? Remove `BUILD_SHARED_LIBS`? Remove something else? :wink:

File: kokkos.presets.json (hunk 5 above)
Comment (romintomasetti) [reply kokkos.presets.json]: Ok, I removed these `"type"` and `"value"` ;)

File: kokkos.presets.json (hunk 4 above)
Comment (romintomasetti) [reply kokkos.presets.json]: @masterleinad

File: CMakePresets.json (hunk 8 above)
Comment (romintomasetti) [reply CMakePresets.json]: Is the current solution with `Kokkos_ROOT` OK for you both @masterleinad and @dalg24 ?

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

File: kokkos.presets.json [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+                "Kokkos_ARCH_VEGA906" : "ON",
+                "CMAKE_CXX_COMPILER"  : "hipcc"
+            }
+        }
+    ],
+    "buildPresets" : [
+        {
+            "name"                        : "OpenMP",
+            "configurePreset"             : "OpenMP",
+            "inheritConfigureEnvironment" : true
```
Comment (dalg24) [inline kokkos.presets.json]: Inheriting an empty environment is kinda pointless.

File: kokkos.presets.json [hunk 18]
```diff
@@ -0,0 +1,58 @@
+{
+    "version"   : 3,
+    "configurePresets" : [
+        {
+            "name"          : "default",
+            "binaryDir"     : "${sourceDir}/build-with-${presetName}",
+            "cacheVariables" : {
+                "CMAKE_EXPORT_COMPILE_COMMANDS" : "ON",
```
Comment (dalg24) [inline kokkos.presets.json]: What is the point of exporting the compile command here?

File: CMakePresets.json [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+    "version"   : 3,
+    "configurePresets" : [
+        {
+            "name"          : "default",
+            "binaryDir"     : "${sourceDir}/build-with-${presetName}",
+            "cacheVariables" : {
+                "CMAKE_EXPORT_COMPILE_COMMANDS"     : "ON",
+                "CMAKE_BUILD_TYPE"                  : "Release",
+                "CMAKE_CXX_STANDARD"                : "17",
+                "KokkosTools_REUSE_KOKKOS_COMPILER" : "ON",
```
Comment (dalg24) [inline CMakePresets.json]: If you are going to use that "reuse cxx compiler option", what is the point of setting the cxx compiler below?

File: kokkos.presets.json (hunk 17 above)
Comment (romintomasetti) [reply kokkos.presets.json]: Indeed, I'll remove that. I guess it was a leftover :smile:

File: kokkos.presets.json (hunk 18 above)
Comment (romintomasetti) [reply kokkos.presets.json]: I guess it was needed mostly for me and myself in some early debugging phase. This isn't needed by default and you can consider it removed :smile:

File: CMakePresets.json (hunk 19 above)
Comment (romintomasetti) [reply CMakePresets.json]: As discussed, we are not even sure that `KokkosTools_REUSE_KOKKOS_COMPILER` works (as it does not set variables with `FORCE` for instance), and as we already set the compiler above it shouldn't be necessary.

File: CMakePresets.json (hunk 8 above)
Comment (romintomasetti) [reply CMakePresets.json]: I guess it's OK for you guys :wink:

**Review by dalg24 — APPROVED** [review]
Thank you
